src/saveload/afterload.cpp

```diff
@@ -3045,6 +3046,40 @@ bool AfterLoadGame()
 		}
 	}
 
+	/* Move ships from lock slope to upper or lower position. */
+	Ship *s;
+	FOR_ALL_SHIPS(s) {
+		/* Suitable tile? */
+		if (!IsTileType(s->tile, MP_WATER) || !IsLock(s->tile) || GetLockPart(s->tile) != LOCK_PART_MIDDLE) continue;
+
+		/* We don't need to adjust position when at the tile centre */
+		int x = s->x_pos & 0xF;
+		int y = s->y_pos & 0xF;
+		if (x == 8 && y == 8) continue;
+
+		/* Test if ship is on the second half of the tile */
+		bool second_half;
+		DiagDirection shipdir = DirToDiagDir(s->direction);
+		switch (shipdir) {
+			default: NOT_REACHED();
+			case DIAGDIR_NE: second_half = x < 8; break;
+			case DIAGDIR_NW: second_half = y < 8; break;
+			case DIAGDIR_SW: second_half = x > 8; break;
+			case DIAGDIR_SE: second_half = y > 8; break;
+		}
```

Comment:
An ugly variant of this switch block is...
second_half = (v->direction == ((x < 8) ? DIR_NE : DIR_SW)) || (v->direction == ((y < 8) ? DIR_NW : DIR_SE))